### PR TITLE
Fetch incident

### PR DESCRIPTION
**Pull request by darkushin**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: demisto/etc/issues/18143

## Description
Adding fetch-incidents option to the integration

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Demisto
- [ ] 4.5.0
- [ ] 5.0.0
- [ ] 5.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation 

## Demisto Partner?
- [ ] The title must be in the following format: **[YOUR_PARTNER_ID] short description**

### CODE REVIEW

File: Packs/MicrosoftGraphSecurity/Integrations/MicrosoftGraphSecurity/MicrosoftGraphSecurity.py [hunk 1]
```diff
@@ -95,6 +95,61 @@ def get_user(self, user_id):
...
+    if incidents:
+        count = 0
+        incidents = sorted(incidents, key=lambda k: k['createdDateTime'])  # sort the incidents by time-increasing order
+        last_incident_time = last_run.get('time', '0')
+        demisto.debug(f'Incidents times: {[incidents[i]["createdDateTime"] for i in range(len(incidents))]}\n')
+        for incident in incidents:
+            incident_time = incident.get('createdDateTime')
+            if incident_time > last_incident_time and count < fetch_limit:
+                demisto_incidents.append({
+                    'name': incident.get('title'),
```
Comment (yuvalbenshalom) [inline Packs/MicrosoftGraphSecurity/Integrations/MicrosoftGraphSecurity/MicrosoftGraphSecurity.py:137]: is title unique to every alert? if so good, if not wi should consider adding an id of the alert to it as the incident name.
also we should consider adding the integration name. lets talk about this.

File: Packs/MicrosoftGraphSecurity/Integrations/MicrosoftGraphSecurity/MicrosoftGraphSecurity.py [hunk 2]
```diff
@@ -95,6 +95,61 @@ def get_user(self, user_id):
...
+                    'severity': severity_map.get(incident.get('severity', ''), 0),  # todo: check with Arseny
+                    'rawJSON': json.dumps(incident)
+                })
+                count += 1
+        if demisto_incidents:
+            last_incident_time = demisto_incidents[-1].get('occurred')
+            new_last_run.update({'time': last_incident_time})
+
+    if not demisto_incidents:
+        new_last_run.update({'time': now})
```
Comment (yuvalbenshalom) [inline Packs/MicrosoftGraphSecurity/Integrations/MicrosoftGraphSecurity/MicrosoftGraphSecurity.py:148]: if you got no demisto incidents, you can keep the last run avoid the race condition that an alert was created just now.(so you can miss it).

File: Packs/MicrosoftGraphSecurity/Integrations/MicrosoftGraphSecurity/MicrosoftGraphSecurity.py [hunk 3]
```diff
@@ -95,6 +95,61 @@ def get_user(self, user_id):
         return response
 
 
+def fetch_incidents(client: MsGraphClient, fetch_time: str, fetch_limit: int, filter_param: str, providers_param: str) \
```
Comment (yuvalbenshalom) [inline Packs/MicrosoftGraphSecurity/Integrations/MicrosoftGraphSecurity/MicrosoftGraphSecurity.py:98]: we should consider calling fetch incidents in test module(without calling last run) when fetch incidents is marked. this wy you will see the errors in the test if there are problems with the fetch configs.

File: Packs/MicrosoftGraphSecurity/Integrations/MicrosoftGraphSecurity/MicrosoftGraphSecurity.yml [hunk 4]
```diff
@@ -21,20 +21,54 @@ configuration:
...
+  display: How many incidents to fetch each time
+  hidden: false
+  name: fetch_limit
+  required: false
+  type: 0
+- display: Fetch incidents of the given providers only. Multiple providers can be
+    inserted, seperated by a comma, e.g. "{first_provider},{second_provioder}". If
+    empty, incidents of all providers will be fetched.
+  hidden: false
+  name: fetch_providers
```
[the remark this answers is not in the document]
Comment (darkushin) [reply Packs/MicrosoftGraphSecurity/Integrations/MicrosoftGraphSecurity/MicrosoftGraphSecurity.yml:60]: If the providers are not valid providers then we won't get any incidents returned

File: Packs/MicrosoftGraphSecurity/Integrations/MicrosoftGraphSecurity/MicrosoftGraphSecurity.py (hunk 1 above)
Comment (darkushin) [reply Packs/MicrosoftGraphSecurity/Integrations/MicrosoftGraphSecurity/MicrosoftGraphSecurity.py:137]: The title isn't unique, but there is an unique ID for every alert. I updated the name of the incident to be 'title + id', for example: 'test alert - da637218501473413212_-1554891308'.
What do you mean by adding the integration name?

File: Packs/MicrosoftGraphSecurity/Integrations/MicrosoftGraphSecurity/MicrosoftGraphSecurity.py (hunk 2 above)
Comment (darkushin) [reply Packs/MicrosoftGraphSecurity/Integrations/MicrosoftGraphSecurity/MicrosoftGraphSecurity.py:148]: I considered it again after our talk. 'now' is a variable that is defined to the current time **before** I call the search_alerts command. The 'time_to' variable of the search alerts is defined according to 'now', so updating the new_last_run according to 'now' is actually the same as updating it with 'time_to', which is the last possible incident that was fetched. Do you still see an option for race condition and missing an alert?

File: Packs/MicrosoftGraphSecurity/Integrations/MicrosoftGraphSecurity/MicrosoftGraphSecurity.py (hunk 3 above)
Comment (darkushin) [reply Packs/MicrosoftGraphSecurity/Integrations/MicrosoftGraphSecurity/MicrosoftGraphSecurity.py:98]: Done. When trying to fetch incidents with an invalid filter format an error is returned as we wanted. Currently I just show the error returned by MS Graph (as was done in other parts in the integration as well), i.e. an error 400 is displayed. Do you want me to handle the error and display a different message or is it informative enough?
image

File: Packs/MicrosoftGraphSecurity/ReleaseNotes/2_0_0.md [hunk 5]
```diff
@@ -0,0 +1,4 @@
+
+#### Integrations
+- __Microsoft Graph__
+Added fetch-incidents function and unit test
```
Comment (amshamah419) [inline Packs/MicrosoftGraphSecurity/ReleaseNotes/2_0_0.md:4]:
```suggestion
  - Added fetch-incidents function and unit test
```

**Comment by kirbles19** [comment]
@darkushin Done.

**Pull request merged**